### PR TITLE
Generate deprecated attribute on deprecated functions

### LINKED ISSUE #517: Set functions deprecated when compilation version >= deprecation version
Something like:

```rust
#[cfg_attr(version = "3.20", deprecated)]
```

should work just fine. Doing it in the next days or so. :)

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #517.

cc @EPashkin @sdroege

### CODE REVIEW

File: src/codegen/general.rs [hunk 1]
```diff
@@ -177,6 +177,36 @@ pub fn define_shared_type(
...
+
+pub fn cfg_deprecated_string(
+    deprecated: Option<Version>,
+    commented: bool,
+    indent: usize,
+) -> Option<String> {
+    if let Some(v) = deprecated {
+        let comment = if commented { "//" } else { "" };
+        Some(format!(
+            "{}{}#[cfg_attr({}, deprecated)]",
```
Comment (sdroege) [inline src/codegen/general.rs]: I don't fully understand how this should work. The optional "deprecated-version" attribute in the gir file is for the native library version, so we would have to e.g. generate a `feature = "v2_40"`. Is that what `v.to_cfg()` does?
What for the case where there is no version, we just ignore that case?

I don't see you parsing the deprecated-version field from the gir file, only something from the toml configuration.

Comment (sdroege) [reply src/codegen/general.rs]: AFAIU it would work as `#[cfg_attr(feature = "v2_40", deprecated)]`, not just `#[cfg_attr("v2_40", deprecated)]`.

Comment (GuillaumeGomez) [reply src/codegen/general.rs]: `to_cfg` generates `feature = "v2_40"`. Also, I added the parse of the deprecated field [here](https://github.com/gtk-rs/gir/pull/518/files#diff-08faf151bfcc6c456f5db4c647591a76R42). But more globally, we already get the deprecated version when parsing the gir files (when I generated for gtk and gio, it got the information). But maybe I'm wrong or misunderstood you?

Comment (sdroege) [reply src/codegen/general.rs]: That line is parsing it from the toml config, not from the gir file. You mean the code to parse it from the gir file was already in src/parser.rs before? All good then

Comment (GuillaumeGomez) [reply src/codegen/general.rs]: Yes, we already got this information before (I didn't add much `deprecated_version` fields as you can see). :)

Comment (sdroege) [reply src/codegen/general.rs]: Lucky you :) All good then

**Comment by GuillaumeGomez** [comment]
Ok ok, I'll update this evening. Thanks for the review!

**Comment by GuillaumeGomez** [comment]
Ok, should be good now. We'll need to add `#![allow(deprecated)]` in gtk-rs non-`sys` crates to avoid the warnings when building them but otherwise that should be all. :)

File: src/codegen/function.rs [hunk 2]
```diff
@@ -49,6 +51,9 @@ pub fn generate(
     let suffix = if only_declaration { ";" } else { " {" };
 
     try!(writeln!(w, ""));
+    if only_declaration {
```
Comment (EPashkin) [inline src/codegen/function.rs]: Need be `!in_trait || only_declaration`, or `g_settings_list_relocatable_schemas` will be not deprecated in gio.

File: src/codegen/general.rs [hunk 3]
```diff
@@ -177,6 +177,40 @@ pub fn define_shared_type(
...
+    Ok(())
+}
+
+pub fn cfg_deprecated_string(
+    deprecated: Option<Version>,
+    env: &Env,
+    commented: bool,
+    indent: usize,
+) -> Option<String> {
+    if env.is_totally_deprecated(deprecated) {
```
Comment (EPashkin) [inline src/codegen/general.rs]: This functions always return false in gdk, due `options.deprecate_by_min_version=false`.
IMHO better add to `Env` function `is_too_low_version`, and write `#[deprecated]` in this case

File: src/codegen/properties.rs [hunk 4]
```diff
@@ -50,6 +50,9 @@ fn generate_prop_func(
     try!(writeln!(w, ""));
 
     let decl = declaration(env, prop);
+    if only_declaration {
```
Comment (EPashkin) [inline src/codegen/properties.rs]: Same as other

File: src/codegen/signal.rs [hunk 5]
```diff
@@ -29,6 +29,9 @@ pub fn generate(
     let suffix = if only_declaration { ";" } else { " {" };
 
     try!(writeln!(w, ""));
+    if only_declaration {
```
Comment (EPashkin) [inline src/codegen/signal.rs]: This too

**Comment by EPashkin** [comment]
Agree about `#![allow(deprecated)]`, then we can even restore deprecation on sys, but IMHO better not do this

**Comment by GuillaumeGomez** [comment]
I prefer not in sys. Let's keep it as "simple" as possible.

**Comment by GuillaumeGomez** [comment]
Updated!

File: src/codegen/signal.rs [hunk 6]
```diff
@@ -73,6 +76,9 @@ pub fn generate(
 
     if let Some(ref emit_name) = analysis.action_emit_name {
         try!(writeln!(w, ""));
+        if only_declaration {
```
Comment (EPashkin) [inline src/codegen/signal.rs]: Maybe this also need "fixed"

File: src/codegen/general.rs [hunk 7]
```diff
@@ -177,6 +177,40 @@ pub fn define_shared_type(
...
+}
+
+pub fn cfg_deprecated_string(
+    deprecated: Option<Version>,
+    env: &Env,
+    commented: bool,
+    indent: usize,
+) -> Option<String> {
+    if env.is_too_low_version(deprecated) {
+        None
```
Comment (EPashkin) [inline src/codegen/general.rs]: Why not return `#[deprecated]` here?

**Comment by EPashkin** [comment]
Code looks good, only last final condition fix,  README.md update for `[[object.member]]`
and maybe "#[deprecated]" generation

**Comment by GuillaumeGomez** [comment]
About the `#[deprecated]` generation, don't you to just keep it outside of the automatic generation? Having it at the top of the crates seems better, no? Instead of having it repeated in every file having some deprecated content generated.

**Comment by EPashkin** [comment]
@GuillaumeGomez You seems mistake with `#![allow(deprecated)]`: agree that it will be added manually.
But if function, for example, deprecated in v3.0, then we better add `#[deprecated]` for it, if `min_cfg_version = "3.4"`

**Comment by GuillaumeGomez** [comment]
Aaaaaaaah! Ok now I got it.

**Comment by EPashkin** [comment]
`analysis::trampolines::Trampoline` don't use `deprecated_version` field and it can be safely removed.

**Comment by GuillaumeGomez** [comment]
Updated!

File: src/codegen/general.rs [hunk 8]
```diff
@@ -177,6 +177,40 @@ pub fn define_shared_type(
...
+}
+
+pub fn cfg_deprecated_string(
+    deprecated: Option<Version>,
+    env: &Env,
+    commented: bool,
+    indent: usize,
+) -> Option<String> {
+    if env.is_too_low_version(deprecated) {
+        Some("#[deprecated]".to_owned())
```
Comment (EPashkin) [inline src/codegen/general.rs]: Comment info and indent will be lost

Comment (GuillaumeGomez) [reply src/codegen/general.rs]: Sorry, quick update on the way!

**Comment by GuillaumeGomez** [comment]
Added missing info and indent.

**Comment by EPashkin** [comment]
Now it ready from my point of view.
@sdroege How it for you?

**Comment by sdroege** [comment]
Seems fine to me but I'd like to look at the output of GIR for e.g. gtk. Is there are PR already?

**Comment by EPashkin** [comment]
As I know there no PR for gtk with this :(

**Comment by EPashkin** [comment]
@sdroege https://github.com/EPashkin/gtk/tree/deprecate with gtk changes

**Comment by sdroege** [comment]
Looks good to me. But do deprecated attributes work in traits? Also are deprecated flags/enum members handled?

**Comment by EPashkin** [comment]
At minimum we know now that deprecated functions only allowed in direct impl and in trait, but not in trait implementation.
I see many warnings about flags and enums, but don't looked at it.

**Comment by GuillaumeGomez** [comment]
Deprecated attributes work in traits *declaration*. For what I tested, they work in enums, didn't check for flags.

**Comment by EPashkin** [comment]
Checked that deprecation works for enum members and trait functions.

**Comment by EPashkin** [comment]
Ok, I merge this.
@GuillaumeGomez Thanks

**Pull request merged**